controllers/secretproviderclasspodstatus_controller.go

```diff
@@ -19,49 +19,136 @@ package controllers
+			}
+		}
+	}
+}
+
+func (r *SecretProviderClassPodStatusReconciler) Patcher() error {
+	log.Debugf("patcher started")
+	r.Mutex.Lock()
+	defer r.Mutex.Unlock()
+
+	ctx := context.Background()
+	spcPodStatusList := &v1alpha1.SecretProviderClassPodStatusList{}
+	spcMap := make(map[string]v1alpha1.SecretProviderClass)
+	secretOwnerMap := make(map[types.NamespacedName][]*v1alpha1.SecretProviderClassPodStatus)
+	// get a list of all spc pod status that belong to the node
+	err := r.Reader.List(ctx, spcPodStatusList, r.ListOptionsLabelSelector())
+	if err != nil {
+		return fmt.Errorf("failed to list secret provider class pod status, err: %+v", err)
+	}
+
+	spcPodStatuses := spcPodStatusList.Items
+	for i := range spcPodStatuses {
+		spcName := spcPodStatuses[i].Status.SecretProviderClassName
+		spc := &v1alpha1.SecretProviderClass{}
+		if val, exists := spcMap[spcPodStatuses[i].Namespace+"/"+spcName]; exists {
+			spc = &val
+		} else {
+			if err := r.Reader.Get(ctx, client.ObjectKey{Namespace: spcPodStatuses[i].Namespace, Name: spcName}, spc); err != nil {
+				log.Errorf("failed to get spc %s, err: %+v", spcName, err)
+				return err
+			}
+			spcMap[spcPodStatuses[i].Namespace+"/"+spcName] = *spc
+		}
+		for _, secret := range spc.Spec.SecretObjects {
+			key := types.NamespacedName{Name: secret.SecretName, Namespace: spcPodStatuses[i].Namespace}
+			val, exists := secretOwnerMap[key]
+			if exists {
+				secretOwnerMap[key] = append(val, &spcPodStatuses[i])
+			} else {
+				secretOwnerMap[key] = []*v1alpha1.SecretProviderClassPodStatus{&spcPodStatuses[i]}
+			}
+		}
+	}
+
+	for secret, owners := range secretOwnerMap {
+		patchFn := func() (bool, error) {
+			if err := r.patchSecretWithOwnerRef(ctx, secret.Name, secret.Namespace, owners...); err != nil {
+				if !apierrors.IsConflict(err) || !apierrors.IsTimeout(err) {
+					log.Errorf("failed to set owner ref for secret, err: %+v", err)
+				}
+				return false, nil
+			}
+			return true, nil
+		}
+		if err := wait.ExponentialBackoff(wait.Backoff{
+			Steps:    5,
+			Duration: 1 * time.Millisecond,
+			Factor:   1.0,
+			Jitter:   0.1,
+		}, patchFn); err != nil {
+			return err
+		}
+	}
+
+	log.Debugf("patcher completed")
+	return nil
+}
+
+// ListOptionsLabelSelector returns a ListOptions with a label selector for node name.
+func (r *SecretProviderClassPodStatusReconciler) ListOptionsLabelSelector() client.ListOption {
+	return client.MatchingLabels(map[string]string{
+		v1alpha1.InternalNodeLabel: r.NodeID,
+	})
+}
+
 // +kubebuilder:rbac:groups=secrets-store.csi.x-k8s.io,resources=secretproviderclasspodstatuses,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=secrets-store.csi.x-k8s.io,resources=secretproviderclasspodstatuses/status,verbs=get;update;patch
 // +kubebuilder:rbac:groups=secrets-store.csi.x-k8s.io,resources=secretproviderclasses,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch
```

Comment:
I think this is only needed when rotation is enabled right? ie store.GetPod is only used by the rotation reconciler.

Reply:
Yeah, this permission is required only for rotation reconciler. We added a separate role for secrets sync feature, but I don't think we want to do the same for pods too?

Reply:
Definitely interested in other's opinions on this. IMO if we are following the principle of least privilege, then this permission on pods is only needed if user wants to support rotation. otherwise, it's not needed.

Reply:
As discussed in last week's community call, "get pod" is acceptable as a base-level rbac for this driver as it does not allow escalation between node workloads.